src/System.Runtime.Extensions/src/System/IO/PathHelper.Windows.cs

```diff
@@ -2,457 +2,373 @@
-
-                Debug.Assert(result < m_maxPath, "did we accidentally remove a PathTooLongException check?");
-                if (result == 0 && m_sb[0] != '\0')
+                else
                 {
-                    if (Length >= m_maxPath)
-                    {
-                        throw new PathTooLongException(SR.IO_PathTooLong);
-                    }
-                    throw Win32Marshal.GetExceptionForLastWin32Error();
+                    return fullPath.ToString();
                 }
-
-                // We have expanded the paths and GetLongPathName may or may not behave differently from before.
-                // We need to call it again to see:
-                doNotTryExpandShortFileName = false;
-
-                m_sb = finalBuffer;
-                return result;
             }
         }
 
-        [System.Security.SecurityCritical]
-        internal unsafe bool TryExpandShortFileName()
+        private static bool IsPossibleShortChar(char value)
         {
-            if (doNotTryExpandShortFileName)
-                return false;
+            // Normal restrictions apply, also cannot be a-z, +, ;, =, [,], 127, above 255, or comma.
+            // Lower case has to be let through as the file system isn't case sensitive and will match against the upper case.
+            return value >= ' '
+                && value <= LastAnsi
+                && value != '+'
+                && value != ';'
+                && value != '='
+                && value != '['
+                && value != ']'
+                && value != ','
+                && value != Delete;
+        }
 
-            if (useStackAlloc)
-            {
-                NullTerminate();
-                char* buffer = UnsafeGetArrayPtr();
-                char* shortFileNameBuffer = stackalloc char[Path.MaxPath + 1];
+        private bool IsUnc(StringBuilder builder)
+        {
+            return builder.Length > 1 && builder[0] == '\\' && builder[1] == '\\';
+        }
 
-                int r = Interop.mincore.GetLongPathNameUnsafe(buffer, shortFileNameBuffer, Path.MaxPath);
+        private StringBuilder GetOutputBuffer(int minimumCapacity)
+        {
+            if (t_cachedOutputBuffer == null)
+            {
+                t_cachedOutputBuffer = new StringBuilder(minimumCapacity);
+            }
+            else
+            {
+                t_cachedOutputBuffer.Clear();
+                t_cachedOutputBuffer.EnsureCapacity(minimumCapacity);
+            }
 
-                // If success, the return buffer length does not account for the terminating null character.
-                // If in-sufficient buffer, the return buffer length does account for the path + the terminating null character.
-                // If failure, the return buffer length is zero 
-                if (r >= Path.MaxPath)
-                    throw new PathTooLongException(SR.IO_PathTooLong);
+            return t_cachedOutputBuffer;
+        }
```

Comment:
It'd be cheaper to write this as:
```C#
StringBuilder local = t_cachedOutputBuffer;
if (local == null)
{
    t_cachedOutputBuffer = local = new StringBuilder(minimumCapacity);
}
else
{
    local.Clear();
    local.EnsureCapacity(minimumCapacity);
}
return local;
```
That said, it'd be better to not even store the builder into t_cachedOutputBuffer until we're done with it and know that it's not too big.